src/lib/salad/hll.c

```diff
@@ -0,0 +1,223 @@
+#include "hll.h"
+#include "hll_empirical.h"
+
+#include <trivia/util.h>
+
+#include <stddef.h>
+#include <math.h>
+#include <assert.h>
+
+/* 6 bits are needed to store the number of leading zeros of 64 bit hash */
+#define HLL_REGISTER_BITS  6
+
+static inline size_t
+hll_n_registers(uint8_t precision)
+{
+	return 1U << precision;
+}
+
+static inline int
+hll_is_valid_cache(double cached)
+{
+	return cached > 0;
+}
+
+static inline void
+hll_invalidate_cache(struct hll *hll)
+{
+	hll->cached_estimation = -1.f;
+}
+
+struct hll *
+hll_create(uint8_t precision)
+{
+	assert(precision >= HLL_MIN_PRECISION);
+	assert(precision <= HLL_MAX_PRECISION);
+	struct hll *hll = xcalloc(1, sizeof(*hll));
+	if (hll) {
+		hll->precision = precision;
+		size_t n_registers = hll_n_registers(precision);
+		size_t register_size = n_registers * HLL_REGISTER_BITS / 8;
+		hll_invalidate_cache(hll);
+		hll->registers = xcalloc(register_size, 1);
+		if (hll->registers == NULL) {
+			free(hll);
+			return NULL;
+		}
+	}
+	return hll;
+}
+
+void
+hll_destroy(struct hll *hll)
+{
+	free(hll->registers);
+	free(hll);
+}
+
+/*
+ * The highest precision bits of the hash are interpreted as a register index.
+ */
+static inline size_t
+hll_hash_register_idx(uint64_t hash, uint8_t precision)
+{
+	return hash >> (64 - precision);
+}
+
+/*
+ * Get a number whose first n bits are equal to ones.
+ */
+static inline uint64_t
+hll_ones(size_t n)
+{
+	return ((1ULL << n) - 1);
+}
+
+/*
+ * Return the number of leading zeros of the first
```

Comment:
I would explain that the zero count is actually the index into the register array — otherwise your assertion seems unobvious.

Reply:
What the assertion?